Add compression.

dellis23 opened the pull request:
👓 @spladug 

I'm erring now on the side of not doing the fork of `gevent-websocket`, just because the way we are doing things here with a single compressor and decompressor means the resulting interface / code will end up being a hack regardless.  Let me know what you think.

Stats stuff is split into another commit just because it's kind of hacky, and we might want to do some of it differently.  I chose to do `percent_compressed` versus something like `bytes_saved` since:

    10,000 -> 1,000 = 9000 bytes saved @ 90% efficiency

While

    100,000 - 91,000 = 9000 bytes saved @ ~9% efficiency

It seems like percentage gives the picture we want.

dellis23 committed:
Add support for compression.

This only adds supports for those clients following RFC 7692 (i.e. those
clients passing and supporting the "permessage-deflate" extension):

    https://tools.ietf.org/html/rfc7692

Those webkit clients using the "x-webkit-deflate-frame" (e.g. safari)
extension are not supported:

    https://tools.ietf.org/id/draft-tyoshino-hybi-websocket-perframe-deflate-05.txt

One global compressor and decompressor is used to save on memory
overhead.  To accomplish this, the server always limits the compression
context to be on a per-frame basis, so context is not needed to be
maintained for each individual client.

dellis23 committed:
Add compression stats.

spladug commented on reddit_service_websockets/dispatcher.py:
```diff
@@ -22,6 +46,14 @@ def __init__(self, metrics):
         consumers = self.consumers.get(namespace, [])
 
+        # Compress the message
+        compressed = make_compressed_frame(message, COMPRESSOR)
```
![KOMPRESSOR](https://upload.wikimedia.org/wikipedia/commons/thumb/1/11/SL55_AMG_kompressor_v8.jpg/220px-SL55_AMG_kompressor_v8.jpg)

spladug replied:
More seriously: a simple optimization is to set a minimum message size before we'll even try compressing. The exact thresholds depend on the data, but could be as low as a couple hundred bytes, or something like 1400 for "eh, it'll fit in a  single internet-MTU packet regardless" depending on the goal. Do you think that'd be worthwhile here to save some CPU in the obviously-wont-help cases?

kjoconnor replied:
We can also test compression and if it comes out larger, just fall back to sending the raw version right?

spladug replied:
Yup, at the cost of server CPU.

kjoconnor replied:
Yeah - I definitely agree about don't bother compressing if it's < MTU, but could also add that size check as well.

spladug replied:
Agreed! No need to send it if we tried and it's worse!

dellis23 replied:
I'm leaning on the side of not doing the "don't bother compressing if it's larger" thing.  If I send the empty payload `{}`, I get a before size of 31 and an after size of 34.  I suspect that by the time we get to the MTU size there's little chance we won't save at least 3 bytes by compressing.

spladug replied:
okay

spladug commented on reddit_service_websockets/dispatcher.py:
```diff
@@ -22,6 +46,14 @@ def __init__(self, metrics):
+        message = Message(
+            compressed=compressed,
+            raw=message,
+            percent_compressed=1 - float(len(compressed))/float(len(message)),
```
Since this is computable downstream given the other two fields in `Message`, it feels a bit odd to compute it up front here far away from the actual stats code that it's for. In other words: why should this code care that the stats code is gonna want this value?

dellis23 replied:
Yeah, my thinking was just that if we do it downstream, the calculation will have to happen once for every time the message is sent out, since there's a pump dispatcher for every client.  I indeed don't like it here either (which is why I put the `max(0, [...])` math elsewhere, but now that I think about it, that's the same problem.  Do you think it's worth it to repeat the calculation for every client or make the compromise on what the Message object is caring about?

spladug replied:
Good point. That actually raises something more serious: sending a metric for each connection is gonna cause a big ol' storm of metrics isn't it?

Could we perhaps aggregate up the savings and send one metric instead?

dellis23 replied:
Hmm, yeah, I guess for live comments / live threads it wont be trivial.  What about a sampling rate?  Or: I'm not sure how we could aggregate and not have smaller namespaces step over the larger ones (for instance, a single orangered namespace might aggregate to a savings of 10%, where a 70k comment thread namespace might aggregate to 50%.  The 50% should completely dwarf the 10% one, but wouldn't if we just sent one metric back per message per namespace).

spladug commented on reddit_service_websockets/patched_websocket.py:
```diff
@@ -0,0 +1,124 @@
+
+def make_compressed_frame(message, compressor):
+    """
+    Generates header and a compressed message which can then be used on any
```
nit: pep257 please. "Generate header..." etc.

spladug commented on reddit_service_websockets/socketserver.py:
```diff
@@ -4,13 +4,33 @@
+
+
+def send_metrics_histogram(metrics_client, name, value):
+    """
```
nit: this is a comment, not a docstring. it should match that syntactically. (i.e. if we ran sphinx on this code, this would not be a useful docstring but it'd be useful when looking at the code itself to understand the whys)

spladug commented on reddit_service_websockets/socketserver.py:
```diff
@@ -111,9 +174,23 @@ def _send_message(self, key, value):
-    def _pump_dispatcher(self, namespace, websocket):
-        for msg in self.dispatcher.listen(namespace, max_timeout=self.ping_interval):
+    def _pump_dispatcher(self, namespace, websocket, supports_compression):
+        for msg in self.dispatcher.listen(
```
bleh, can't this just go a bit wider and not have the ugly extra indent? :P

spladug commented on reddit_service_websockets/socketserver.py:
```diff
@@ -61,8 +81,47 @@ def upgrade_connection(self):
+        # Ideally, this would probably go in `upgrade_connection`, but we have
+        # no way to modify the headers there without duplicating the logic of
+        # the entire method.  This is a much smaller entry point.
+        # Ideally, geventwebsocket would just support this stuff.
```
since we've got monkeypatches already, would this be something we could hack into it?

dellis23 replied:
I guess in retrospect I kind of like keeping it here, since it keeps the header negotiation stuff in one module and the patched module doesn't need to know about it.  Since we're adding this notion of putting `supports_compression` in the environment so that the socket server can check it, if we were to move it into that patched module, _both_ modules would need to be aware of it.  Does that seem reasonable?  I can also update the comment given that my view has changed.

spladug replied:
ok, updating comment seems reasonable

spladug reviewed (changes requested):
I think I'm fine with the patching we're doing here rather than modifying upstream. This looks good overall. A few little nits and a question about using size to skip compression are all I've got. Nice!

dellis23 committed:
Don't indent at expense of 80 char limit.

Mr. PEP8 would roll in his grave.

dellis23 committed:
Change docstring to comment.

dellis23 committed:
Make docstring PEP 257 friendly.

dellis23 committed:
Remove unused imports.

dellis23 committed:
Only compress if message > MTU - headers.

dellis23 commented:
:haircut_man:

dellis23 commented:
Oh, uh, not quite sure what to do about the compression rate metrics, given the good point that it would be very spikey on graphite.  I might just remove it for now until we come up with a better idea.

spladug commented on reddit_service_websockets/dispatcher.py:
```diff
@@ -26,6 +26,21 @@
+# 1440 - common MTU size
+# 60 - TCP max header size
+# 60 - IP max header size
+MIN_COMPRESS_SIZE = 1400 - 60 - 60
```
comment says 1440 but this is 1400?

dellis23 replied:
Amazing that the docs drifted from the code so fast...

spladug commented on reddit_service_websockets/dispatcher.py:
```diff
@@ -47,11 +62,13 @@ def on_message_received(self, namespace, message):
 
         # Compress the message
-        compressed = make_compressed_frame(message, COMPRESSOR)
+        compressed = make_compressed_frame(message, COMPRESSOR) \
```
this feels pretty awkward (expression-if split across lines). it would be more obvious just written as a proper statement-style if.

dellis23 replied:
I think it's beautiful :P  But yeah, I'll change it.

spladug reviewed (changes requested):
Hm, yeah, now that I think more about it, let's just drop the compression metrics. We'll see the change in bandwidth and that's the real metric that we care about. Sound OK? Otherwise, this looks great and my only other complaints are tiny nits.

dellis23 committed:
Fix MTU size and comments.

dellis23 committed:
Break into multiple lines.

dellis23 committed:
Partially revert "Add compression stats."

dellis23 committed:
Remove obsolete comment.

dellis23 commented:
:haircut_man: Yay, no more histogram / timer hack.

spladug reviewed (approved):
delightful!

dellis23 commented:
09b85770692e7d827ed7efaa668943f4812f03d1

dellis23 commented:
Something was up with this on deploy:

```
Jan 25 11:05:23 ws-015746bfd997ac7fe [websockets] Traceback (most recent call last):#015
Jan 25 11:05:23 ws-015746bfd997ac7fe [websockets]  File "/usr/lib/python2.7/dist-packages/gevent/greenlet.py", line 327, in run#015
Jan 25 11:05:23 ws-015746bfd997ac7fe [websockets]    result = self._run(*self.args, **self.kwargs)#015
Jan 25 11:05:23 ws-015746bfd997ac7fe [websockets]  File "/usr/lib/python2.7/dist-packages/gevent/pywsgi.py", line 658, in handle#015
Jan 25 11:05:23 ws-015746bfd997ac7fe [websockets]    handler.handle()#015
Jan 25 11:05:23 ws-015746bfd997ac7fe [websockets]  File "/usr/lib/python2.7/dist-packages/gevent/pywsgi.py", line 184, in handle#015
Jan 25 11:05:23 ws-015746bfd997ac7fe [websockets]    result = self.handle_one_request()#015
Jan 25 11:05:23 ws-015746bfd997ac7fe [websockets]  File "/usr/lib/python2.7/dist-packages/gevent/pywsgi.py", line 321, in handle_one_request#015
Jan 25 11:05:23 ws-015746bfd997ac7fe [websockets]    self.handle_one_response()#015
Jan 25 11:05:23 ws-015746bfd997ac7fe [websockets]  File "/usr/lib/python2.7/dist-packages/gevent/pywsgi.py", line 515, in handle_one_response#015
Jan 25 11:05:23 ws-015746bfd997ac7fe [websockets]    self.handle_error(*sys.exc_info())#015
Jan 25 11:05:23 ws-015746bfd997ac7fe [websockets]  File "/usr/lib/python2.7/dist-packages/gevent/pywsgi.py", line 527, in handle_error#015
Jan 25 11:05:23 ws-015746bfd997ac7fe [websockets]    self.start_response(_INTERNAL_ERROR_STATUS, _INTERNAL_ERROR_HEADERS[:])#015
Jan 25 11:05:23 ws-015746bfd997ac7fe [websockets]  File "/opt/websockets/reddit_service_websockets/socketserver.py", line 88, in start_response#015
Jan 25 11:05:23 ws-015746bfd997ac7fe [websockets]    if self.environ["supports_compression"]:#015
Jan 25 11:05:23 ws-015746bfd997ac7fe [websockets] KeyError: 'supports_compression'#015
```

dellis23 committed:
Always set support_compression.

Do a safe get from the environment, since we fall back to not
compressing, which is safe.

dellis23 commented:
👓 @spladug

dellis23 commented:
6a9504fd63d89802cd98e4a63ad10d90bba031f4
